mijia/src/lib.rs

```diff
@@ -192,36 +231,133 @@ impl MijiaSession {
+        let value = self
+            .bt_session
+            .read_characteristic_value(id, HISTORY_LAST_RECORD_CHARACTERISTIC_PATH)
+            .await?;
+        Ok(Record::decode(&value)?)
+    }
+
+    /// Start receiving historical records from the sensor.
+    ///
+    /// # Arguments
+    /// * `id`: The ID of the sensor to request records from.
+    /// * `start_index`: The record index to start at. If this is not specified then all records
+    ///   which have not yet been received from the sensor since it was connected will be requested.
+    pub async fn start_notify_history(
+        &self,
+        id: &DeviceId,
+        start_index: Option<u32>,
+    ) -> Result<(), BluetoothError> {
+        if let Some(start_index) = start_index {
+            self.bt_session
+                .write_characteristic_value(
+                    id,
+                    HISTORY_INDEX_CHARACTERISTIC_PATH,
+                    start_index.to_le_bytes(),
+                )
+                .await?
+        }
+        self.bt_session
+            .start_notify(id, HISTORY_RECORDS_CHARACTERISTIC_PATH)
+            .await
+    }
+
+    /// Stop receiving historical records from the sensor.
+    pub async fn stop_notify_history(&self, id: &DeviceId) -> Result<(), BluetoothError> {
+        self.bt_session
+            .stop_notify(id, HISTORY_RECORDS_CHARACTERISTIC_PATH)
+            .await
+    }
+
+    /// Try to get all historical records for the sensor.
+    pub async fn get_all_history(&self, id: &DeviceId) -> Result<Vec<Option<Record>>, MijiaError> {
+        let history_range = self.get_history_range(&id).await?;
+        // TODO: Get event stream that is filtered by D-Bus.
+        let (msg_match, events) = self.event_stream().await?;
+        let mut events = events.timeout(HISTORY_RECORD_TIMEOUT);
+        self.start_notify_history(&id, Some(0)).await?;
+
+        let mut history = vec![None; history_range.len()];
+        while let Some(Ok(event)) = events.next().await {
+            match event {
+                MijiaEvent::HistoryRecord {
+                    id: record_id,
+                    record,
+                } => {
+                    log::trace!("{:?}: {}", record_id, record);
+                    if record_id == *id {
+                        if history_range.contains(&record.index) {
+                            let offset = record.index - history_range.start;
+                            history[offset as usize] = Some(record);
+                        } else {
+                            log::error!(
+                                "Got record {:?} for sensor {:?} out of bounds {:?}",
+                                record,
+                                id,
+                                history_range
+                            );
+                        }
+                    } else {
+                        log::warn!("Got record for wrong sensor {:?}", record_id);
```

Comment:
I expect these to be quite noisy in some cases. Probably fine for now though.

Reply:
That should only happen if we try to retrieve history for two sensors simultaneously, and it will be fixed once we do the TODO above to filter the D-Bus stream.